pennylane_qulacs/qulacs_device.py

```diff
@@ -39,181 +74,240 @@
+
+                if op.name in ["QubitUnitary.inv", "Hermitian.inv"]:
+                    par[0] = par[0].conj().T
+
+                # reverse wires (could also change par[0])
+                unitary_gate = gate.DenseMatrix(wires[::-1], par[0])
+                self._circuit.add_gate(unitary_gate)
+                unitary_gate.update_quantum_state(self._state)
+
+            elif op.name in ["Rot", "Rot.inv"]:
+                if len(wires) != 1:
+                    raise ValueError("Rotation gate can only be applied on a single wire.")
 
-    def _get_operator_matrix(self, operation, par):
-        A = self._observable_map[operation]
-        if not callable(A):
-            return A
+                if op.name == "Rot.inv":
+                    par = par[::-1]
+                else:
+                    # Negating the parameters such that it adheres to qulacs
+                    par = np.negative(op.parameters)
 
-        return A(*par)
+                self._circuit.add_gate(gate.RZ(wires[0], par[0]))
+                gate.RZ(wires[0], par[0]).update_quantum_state(self._state)
+                self._circuit.add_gate(gate.RY(wires[0], par[1]))
+                gate.RY(wires[0], par[1]).update_quantum_state(self._state)
+                self._circuit.add_gate(gate.RZ(wires[0], par[2]))
+                gate.RZ(wires[0], par[2]).update_quantum_state(self._state)
 
-    def _get_tensor_operator_matrix(self, obs, par):
-        ops = [self._get_operator_matrix(o, p) for o, p in zip(obs, par)]
-        return functools.reduce(np.kron, ops)
+            elif op.name in ["CRZ", "CRZ.inv", "PhaseShift", "PhaseShift.inv"]:
+                if callable(mapped_operation):
+                    gate_matrix = mapped_operation(*par)
+                else:
+                    gate_matrix = mapped_operation
+
+                # gate_matrix is already in correct order => no wire-reversal needed
+                dense_gate = gate.DenseMatrix(wires, gate_matrix)
+                self._circuit.add_gate(dense_gate)
+                gate.DenseMatrix(wires, gate_matrix).update_quantum_state(self._state)
+
+            else:
+                # Negating the parameters such that it adheres to qulacs
+                par = np.negative(op.parameters)
+
+                # mapped_operation is already in correct order => no wire-reversal needed
+                self._circuit.add_gate(mapped_operation(*wires, *par))
+                mapped_operation(*wires, *par).update_quantum_state(self._state)
+
+    @staticmethod
+    def _get_inverse_operation(mapped_operation, wires, par):
+        """Return the inverse of an operation"""
+        if mapped_operation is None:
+            return mapped_operation
+
+        # if an inverse variant of the operation exists
+        inverse_operation = getattr(gate, mapped_operation.get_name() + "dag", None)
+        if inverse_operation is None:
```

Comment:
Going back to how it was implemented earlier (since the change didn't work).
```suggestion
        try:
            inverse_operation = getattr(gate, mapped_operation.get_name() + "dag")
        except AttributeError:
```